### DISCUSSION TITLE
New 'generator' proposal

### DISCUSSION THREAD

**Discussion by Moki26** [networkx/networkx#7663] [Ideas]
Hello networkx developers,

I use networkx to study various parameters of brain networks and wanted to ask for your opinion on a new graph 'generator' (similar to the social network graphs), which generates graphs representing real anatomical connectivity of human brain areas.

I could develop the relevant python script (according to your guidelines) to be included as a part of the networkx generator set. Technically, it would return an nxn matrix representing connectivity strengths between standardized brain areas (e.g. n=45 for the "Automated anatomical labeling atlas" of one brain hemisphere) taken from published literature.

I think this can be useful as another simple example of a real world network that other users can play with.
Would you find the inclusion of such a generator function an acceptable addition to networkx?

**Comment by dschult** [comment]
Would this be a specific network based on published literature (like the Petersen graph)? Or are you talking about many such networks? How would we choose which get provided? There are so many ways to represent connectivity strengths, its not clear that there is a consensus in the literature for what network representation to use.

Or are we talking about a random network generator with properties that are similar to regions of brain networks? Again, it would be hard to find a consensus in the literature for what that would be.  We don't want to become a database of networks. But if there are "famous examples" which have helped us validate algorithms, like the social network examples we provide for "community detection", we would consider including them. Are there small brain networks that have been studied with respect to illuminating specific algorithms?

Our loose [policy on new functionality](https://output.circle-artifacts.com/output/job/0d498818-05c0-4ed6-8679-2ad8ac785cb6/artifacts/0/doc/build/html/developer/new_contributor_faq.html#q-what-is-the-policy-for-deciding-whether-to-include-a-new-algorithm) is that it should be based on a body of literature (it can be a fairly small body, but more than an idea from a paper).

**Reply by Moki26** [reply]
> Would this be a specific network based on published literature (like the Petersen graph)? 

It would be a single network, representing the 'structural connectivity' of the brain, which is highly conserved across individuals, so in this sense we would not have to choose from many different connectivities. What one would have to decide on is the 'parcellation' (essentially how detailed it is/how many areas the brain is divided into). The suggested AAL parcelation is a standard and a coarse one, but there are indeed other parcellations. I believe AAL's 45x45 size (or 90x90 for both hemispheres) is reasonable so that the generator file remain small enough.

When it comes to connectivity representation, it is fairly common in neuro literature to just use a 2D matrix with artificial unit strengths, so the same was proposed here.

> Our loose policy on new functionality is that it should be based on a body of literature (it can be a fairly small body, but more than an idea from a paper).

Studying graph measures, esp. communities, in the human connectome is quite a body of literature, so I believe having a representative would be nice. The proposed option could be based on data published in https://doi.org/10.1038/s41597-022-01596-9. Unlike other examples, it has a well documented pipeline with fully processed data and it shows consitency with other published connectivities.

I understand that the goal is not for networkx to be a database of networks, so let me know.

**Reply by dschult** [reply]
Hello @moki26 
Sorry for the delay -- this topic brought up other topics and we took a while to find an answer.
The answer is that we do not want to add generators that are based on domain specific measured datasets. We do want to include algorithmic ways to construct networks as generators. Of course, as you know, we do have a couple domain specific datasets which have been included in the package since its inception. We do not expect to get rid of those, but we aren't particularly happy about them being here either. We are trusting that other orgs will provide collections of data representing networks.

I think the distinction is, if a network is created from measured data we want folks to obtain that data from a data repository and bring it to NetworkX. If a network is constructed algorithmically, that belongs in our generator collection. And we also welcome algorithms which take data and turn it into a network. 

Thank you for your interest and for asking a great question which helped us define who we are a little bit. :)
